### ISSUE TITLE
duplicate symbols in libActon.a

### ISSUE THREAD

**Issue by sydow** [actonlang/acton#675]
When compiling programs importing both numpy and math things go wrong:

```
acton> actonc examples/loess_smoother5.act --root main
Compiling loess_smoother5.act for release
ERROR: internal compiler error: compilation of generated C code of the root actor failed
NOTE: this is likely a bug in actonc, please report this at:
NOTE: https://github.com/actonlang/acton/issues/new?template=ice.md
NOTE: acton 0.10.0.20220610.15.44.19 compiled by ghc 8.10 on darwin x86_64
NOTE: cc: Apple clang version 11.0.0 (clang-1100.0.33.17)
cc stdout:

cc stderr:
duplicate symbol '_math$$RealFuns$__init__' in:
    /Users/sydow/acton/dist/lib/rel/libActon.a(math.o)
    /Users/sydow/acton/dist/lib/rel/libActon.a(numpy.o)
duplicate symbol '_math$$RealFuns$float$__deserialize__' in:
    /Users/sydow/acton/dist/lib/rel/libActon.a(math.o)
    /Users/sydow/acton/dist/lib/rel/libActon.a(numpy.o)
duplicate symbol '_math$$RealFuns$float$__init__' in:
    /Users/sydow/acton/dist/lib/rel/libActon.a(math.o)
    /Users/sydow/acton/dist/lib/rel/libActon.a(numpy.o)
``` 

It goes on to list all the symbols in math. Apparently numpy.o contains also math.o and both are part of libActon.a.

Looking at stdlib/Makefile one finds from line 24 
```
NUMPY_CFILES=$(wildcard c_src/numpy/*.h)
ifeq ($(shell uname -s),Linux)
NUMPY_CFLAGS+=-lbsd -ldl -lmd
endif
out/dev/lib/numpy.o: src/numpy.c src/numpy.h out/types/math.h $(NUMPY_CFILES) out/dev/lib/math.o
	@mkdir -p $(dir $@)
	$(CC) $(CFLAGS) $(CFLAGS_DEV) -Wno-unused-result -r -Iout/ $< -o$@ $(NUMPY_CFLAGS) out/dev/lib/math.o

out/rel/lib/numpy.o: src/numpy.c src/numpy.h out/types/math.h $(NUMPY_CFILES) out/rel/lib/math.o
	@mkdir -p $(dir $@)
	$(CC) $(CFLAGS) $(CFLAGS_REL) -Wno-unused-result -r -Iout/ $< -o$@ $(NUMPY_CFLAGS) out/rel/lib/math.o
``` 

We note that math.o is given as argument to CC here; I removed it and did make clean; make and things seem to work.

I also found it strange that in the def of NUMPY_CFILES file suffix .h is used. I also changed this to .c.

So, even if it seems that I solved some problem I certainly do not have a full understanding...

**Comment by sydow** [comment]
Sorry for the unreadable output; obviously some characters from the makefile caused some strange formatting.

The questions are about the Makefile in acton/stdlib:
- why is the file suffix .h used on line 24 and not .c?
- why is the last argument out/XXX/lib/math.o given to CC on lines 30 and 34?
Removing these and changing suffix to .c and rebuilding removes the problem indicated in the beginning of the previous comment.

**Comment by plajjan** [comment]
@sydow don't have the answer to your questions on top of my head. I have written pretty much the entire Makefile so I'm surely at fault. I agree that it doesn't seem to make much sense... but I have a vague recollection of adding math.o for some reason.

Please open a PR with your changes and I'll review it + try to dig a little into history to figure out why it might look like this, but by the sound of it, your suggestions are correct.

**Comment by plajjan** [comment]
Also clearly shows how our tests do not cover this area, so perhaps it's time to run some simple tests for numpy too...

**Comment by plajjan** [comment]
Commit adding math.o: https://github.com/actonlang/acton/commit/8585a5bb7ffa5b9337b6ae57fd645536d1e33842

**Comment by plajjan** [comment]
`NUMPY_CFILES=$(wildcard c_src/numpy/*.h)` is wrong but the variable `NUMPY_CFILES` is only used to specify dependencies of this target, so it doesn't actually affect the output of the target. It only leads to invalid makefiles dependencies, i.e. make will not understand it should rebuild this target if one of the .c files has changed. It will inspects its dependencies and think the target is up to date.

Should be fixed of course, just not the underlying problem of this issue.

**Comment by plajjan** [comment]
So the problem I think I tried to fix here is that I want to compartmentalize modules. numpy requires multiple libraries on Linux:
```
NUMPY_CFLAGS+=-lbsd -ldl -lmd
```

To get this to work, actonc will need to link all built Acton programs with these libraries (well, the linker will naturally not link them if they are not required). actonc doesn't understand C, so it couldn't possibly determine dynamically what libraries need to be included when, we must tell it.

I want to avoid having a hardcoded list in actonc of libraries we need to link. If numpy or any other module (maybe not even in stdlib) needs to be linked with something new, it is undesirable to require changing & upgrading actonc. Thus, my attempt here was to hide numpy's dependencies by partially linking (option `-r` versus `-c` which stops cc after compilation)... my guess is that when I did that partial linking, I also had to include the math module... so that's why math symbols end up in numpy and we then get the duplicate.

Now, this happens for numpy but is a generic problem so we need a solution. I reckon we'd get similar duplicate errors if another module depended on, say `-lbsd`.

I don't know what the solution is though. Maybe go back to a hardcoded list of libs in actonc for the time being!?

**Comment by plajjan** [comment]
There is a `--allow-multiple-definition` option to `ld`. I think it should be safe for us to use that since we know the duplicates really are the same...

**Comment by plajjan** [comment]
If we do not link numpy with bsd/dl/md at compile time we either need to pack those libs into libActon or they will become run time dependencies of actonc.

**Comment by plajjan** [comment]
In #688 I tried not linking numpy with math.o, which works great on Linux but clang on Mac spits out errors... not sure how to fix...

I also did try --allow-multiple-definition, which also works on Linux, but not on mac...

**Comment by plajjan** [comment]
Fixed through #693. Closing.

`-fno-common` was the magic incantation necessary to fix this on MacOS / clang!
